File: xwords4/android/app/src/main/java/org/eehouse/android/xw4/Toolbar.java
/* -*- compile-command: "find-and-gradle.sh inXw4dDeb"; -*- */
/*
 * Copyright 2009-2014 by Eric House (xwords@eehouse.org).  All rights
 * reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

package org.eehouse.android.xw4;

import android.app.Activity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.HorizontalScrollView;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.ScrollView;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;


import org.eehouse.android.xw4.DlgDelegate.Action;
import org.eehouse.android.xw4.DlgDelegate.HasDlgDelegate;
import org.eehouse.android.xw4.loc.LocUtils;

public class Toolbar implements BoardContainer.SizeChangeListener {
    private static final String TAG = Toolbar.class.getSimpleName();
    public enum Buttons {
        BUTTON_HINT_PREV(R.id.prevhint_button),
        BUTTON_HINT_NEXT(R.id.nexthint_button),
        BUTTON_CHAT(R.id.chat_button, R.string.key_disable_chat_button),
        BUTTON_JUGGLE(R.id.shuffle_button, R.string.key_disable_shuffle_button),
        BUTTON_UNDO(R.id.undo_button, R.string.key_disable_undo_button),
        BUTTON_BROWSE_DICT(R.id.dictlist_button, R.string.key_disable_dicts_button),
        BUTTON_VALUES(R.id.values_button, R.string.key_disable_values_button),
        BUTTON_FLIP(R.id.flip_button, R.string.key_disable_flip_button),
        ;

        private int mId;
        private int mKey;
        private Buttons(int id) { this( id, 0); }
        private Buttons(int id, int key) {
            mId = id;
            mKey = key;
        }
        public int getResId() { return mId; }
        public int getDisableId() { return mKey; }
    };

    private Activity m_activity;
    private DlgDelegate.HasDlgDelegate m_dlgDlgt;
    private LinearLayout m_layout;
    private boolean m_visible;
    private Map<Buttons, Object> m_onClickListeners = new HashMap<>();
    private Map<Buttons, Object> m_onLongClickListeners = new HashMap<>();
    private Set<Buttons> m_enabled = new HashSet<>();

    public Toolbar( Activity activity, HasDlgDelegate dlgDlgt )
    {
        m_activity = activity;
        m_dlgDlgt = dlgDlgt;

        BoardContainer.registerSizeChangeListener( this );
    }

    public void setVisible( boolean visible )
    {
        if ( m_visible != visible ) {
            m_visible = visible;
            doShowHide();
        }
    }

    public ImageButton getButtonFor( Buttons index )
    {
        return (ImageButton)m_activity.findViewById( index.getResId() );
    }

    public Toolbar setListener( Buttons index, final int msgID,
                                final int prefsKey, final Action action )
    {
        m_onClickListeners.put( index, new View.OnClickListener() {
                @Override
                public void onClick( View view ) {
                    // Log.i( TAG, "setListener(): click on %s with action %s",
                    //        view.toString(), action.toString() );
                    m_dlgDlgt.makeNotAgainBuilder( prefsKey, action, msgID )
                        .show();
                }
            } );
        return this;
    }

    public Toolbar setLongClickListener( Buttons index, final int msgID,
                                         final int prefsKey, final Action action )
    {
        m_onLongClickListeners.put( index, new View.OnLongClickListener() {
                public boolean onLongClick( View view ) {
                    m_dlgDlgt.makeNotAgainBuilder( prefsKey, action, msgID )
                        .show();
                    return true;
                }
            } );
        return this;
    }

    public Toolbar update( Buttons index, boolean enable )
    {
        if ( enable ) {
            int disableKeyID = index.getDisableId();
            if ( 0 != disableKeyID ) {
                enable = !XWPrefs.getPrefsBoolean( m_activity, disableKeyID,
                                                   false );
            }
        }

        int id = index.getResId();
        ImageButton button = (ImageButton)m_activity.findViewById( id );
        if ( null != button ) {
            button.setVisibility( enable ? View.VISIBLE : View.GONE );
        }

        if ( enable ) {
            m_enabled.add( index );
        } else {
            m_enabled.remove( index );
        }
        return this;
    }

    protected int enabledCount() { return m_enabled.size(); }

    // SizeChangeListener
    @Override
    public void sizeChanged( int width, int height, boolean isPortrait )
    {
        installListeners();
        doShowHide();
    }

    public void installListeners()
    {
        tryAddListeners( m_onClickListeners );
        tryAddListeners( m_onLongClickListeners );
    }

    private void tryAddListeners( Map<Buttons, Object> map )
    {
        Iterator<Buttons> iter = map.keySet().iterator();
        while ( iter.hasNext() ) {
            Buttons key = iter.next();
            Object listener = map.get( key );
            if ( setListener( key, listener ) ) {
                iter.remove();
            }
        }
    }

    private boolean setListener( Buttons index, Object listener )
    {
        ImageButton button = getButtonFor( index );
        boolean success = null != button;
        if ( success ) {
            if ( listener instanceof View.OnClickListener ) {
                button.setOnClickListener( (View.OnClickListener)listener );
            } else if ( listener instanceof View.OnLongClickListener ) {
                button.setOnLongClickListener( (View.OnLongClickListener)listener );
            } else {
                Assert.failDbg();
            }
        }
        return success;
    }

    private void doShowHide()
    {
        boolean isPortrait = BoardContainer.getIsPortrait();

        if ( null == m_layout ) {
            m_layout = (LinearLayout)LocUtils.inflate( m_activity, R.layout.toolbar );
        } else {
            ((ViewGroup)m_layout.getParent()).removeView( m_layout );
        }
        m_layout.setOrientation( isPortrait ?
                                 LinearLayout.HORIZONTAL : LinearLayout.VERTICAL );

        int scrollerId = isPortrait ? R.id.tbar_parent_hor : R.id.tbar_parent_vert;
        ViewGroup scroller = (ViewGroup)m_activity.findViewById( scrollerId );
        if ( null != scroller ) {
            // Google's had reports of a crash adding second view
            scroller.removeAllViews();
            scroller.addView( m_layout );

            scroller.setVisibility( m_visible? View.VISIBLE : View.GONE );
        }
    }
}

File: xwords4/android/app/src/main/java/org/eehouse/android/xw4/Toolbar.kt
/*
 * Copyright 2009 - 2024 by Eric House (xwords@eehouse.org).  All rights
 * reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */
package org.eehouse.android.xw4

import android.app.Activity
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.LinearLayout

import java.util.EnumMap

import org.eehouse.android.xw4.BoardContainer.SizeChangeListener
import org.eehouse.android.xw4.DlgDelegate.Action
import org.eehouse.android.xw4.DlgDelegate.HasDlgDelegate
import org.eehouse.android.xw4.loc.LocUtils

private val TAG: String = Toolbar::class.java.simpleName

class Toolbar(private val mActivity: Activity, private val mDlgDlgt: HasDlgDelegate) :
    SizeChangeListener
{
    enum class Buttons(val resId: Int, val disableId: Int = 0) {
        BUTTON_HINT_PREV(R.id.prevhint_button),
        BUTTON_HINT_NEXT(R.id.nexthint_button),
        BUTTON_CHAT(R.id.chat_button, R.string.key_disable_chat_button),
        BUTTON_JUGGLE(R.id.shuffle_button, R.string.key_disable_shuffle_button),
        BUTTON_UNDO(R.id.undo_button, R.string.key_disable_undo_button),
        BUTTON_BROWSE_DICT(R.id.dictlist_button, R.string.key_disable_dicts_button),
        BUTTON_VALUES(R.id.values_button, R.string.key_disable_values_button),
        BUTTON_FLIP(R.id.flip_button, R.string.key_disable_flip_button),
    }

    private var mLayout: LinearLayout? = null
    private var mVisible = false
    private val mOnclicklisteners: MutableMap<Buttons, Any>
        = EnumMap(Buttons::class.java)
    private val mOnlongclicklisteners: MutableMap<Buttons, Any>
        = EnumMap(Buttons::class.java)
    private val mEnabled: MutableSet<Buttons> = HashSet()

    init {
        BoardContainer.registerSizeChangeListener(this)
    }

    fun setVisible(visible: Boolean) {
        if (mVisible != visible) {
            mVisible = visible
            doShowHide()
        }
    }

    fun getButtonFor(index: Buttons): ImageButton {
        return mActivity.findViewById<View>(index.resId) as ImageButton
    }

    fun setListener(index: Buttons, msgID: Int,
                    prefsKey: Int, action: Action): Toolbar
    {
        mOnclicklisteners[index] =
            View.OnClickListener {
                // Log.i( TAG, "setListener(): click on %s with action %s",
                //        view.toString(), action.toString() );
                mDlgDlgt.makeNotAgainBuilder(prefsKey, action, msgID)
                    .show()
            }
        return this
    }

    fun setLongClickListener(
        index: Buttons, msgID: Int,
        prefsKey: Int, action: Action
    ): Toolbar {
        mOnlongclicklisteners[index] = OnLongClickListener {
            mDlgDlgt.makeNotAgainBuilder(prefsKey, action, msgID)
                .show()
            true
        }
        return this
    }

    fun update(index: Buttons, enable: Boolean): Toolbar {
        var enable = enable
        if (enable) {
            val disableKeyID = index.disableId
            if (0 != disableKeyID) {
                enable = !XWPrefs.getPrefsBoolean(
                    mActivity, disableKeyID,
                    false
                )
            }
        }

        val button = mActivity.findViewById<View>(index.resId) as ImageButton?
        button?.let {
            it.visibility = if (enable) View.VISIBLE else View.GONE
        }

        if (enable) {
            mEnabled.add(index)
        } else {
            mEnabled.remove(index)
        }
        return this
    }

    fun enabledCount(): Int {
        return mEnabled.size
    }

    // SizeChangeListener
    override fun sizeChanged(width: Int, height: Int, isPortrait: Boolean) {
        installListeners()
        doShowHide()
    }

    fun installListeners() {
        tryAddListeners(mOnclicklisteners)
        tryAddListeners(mOnlongclicklisteners)
    }

    private fun tryAddListeners(map: MutableMap<Buttons, Any>) {
        val iter = map.keys.iterator()
        while (iter.hasNext()) {
            val key = iter.next()
            val listener = map[key]
            setListener(key, listener)
            iter.remove()
        }
    }

    private fun setListener(index: Buttons, listener: Any?) {
        val button = getButtonFor(index)
        when (listener) {
            is View.OnClickListener ->
                button.setOnClickListener(listener)
            is OnLongClickListener ->
                button.setOnLongClickListener(listener)
            else -> Assert.failDbg()
        }
    }

    private fun doShowHide() {
        val isPortrait = BoardContainer.getIsPortrait()

        if (null == mLayout) {
            mLayout = LocUtils.inflate(mActivity, R.layout.toolbar) as LinearLayout
        } else {
            (mLayout!!.parent as ViewGroup).removeView(mLayout)
        }
        mLayout!!.orientation = if (isPortrait) LinearLayout.HORIZONTAL else LinearLayout.VERTICAL

        val scrollerId = if (isPortrait) R.id.tbar_parent_hor else R.id.tbar_parent_vert
        val scroller = mActivity.findViewById<ViewGroup>(scrollerId)
        // Google's had reports of a crash adding second view
        scroller.removeAllViews()
        scroller.addView(mLayout)

        scroller.visibility = if (mVisible) View.VISIBLE else View.GONE
    }
}
